napari/layers/utils/text.py

```diff
@@ -39,74 +43,68 @@ class TextManager:
-        Should be 'center', 'upper_left', 'upper_right', 'lower_left', or 'lower_right'.
-    translation : np.ndarray
-        Offset from the anchor point.
-    color : array
-        Font color for the text
-    size : float
-        Font size of the text. Default value is 12.
-    The blending mode that determines how RGB and
-        alpha values of the layer visual get mixed. Allowed values are
-        {'opaque', 'translucent', and 'additive'}. Note that 'opaque` blending
-        is not recommended, as colors the bounding box surrounding the text.
-    visible : bool
-        Set to true of the text should be displayed.
     """
 
-    def __init__(
-        self,
-        text,
-        n_text,
-        properties={},
-        rotation=0,
-        translation=0,
-        anchor='center',
-        color='cyan',
-        size=12,
-        blending='translucent',
-        visible=True,
-    ):
-
-        self.events = EmitterGroup(
-            source=self,
-            auto_connect=True,
-            text=Event,
-            rotation=Event,
-            translation=Event,
-            anchor=Event,
-            color=Event,
-            size=Event,
-            blending=Event,
-            visible=Event,
-        )
-
-        self.events.block_all()
-        self._rotation = rotation
-        self._anchor = Anchor(anchor)
-        self._translation = translation
-        self._color = transform_color(color)[0]
-        self._size = size
-        self._blending = self._check_blending_mode(blending)
-        self._visible = visible
-
-        self._set_text(text, n_text, properties)
-        self.events.unblock_all()
+    text: Union[np.ndarray, str]
+    n_text: int
+    properties: Dict[str, np.ndarray] = field(default_factory=dict)
+    rotation: float = 0.0
+    translation: np.ndarray = np.asarray(0)
+    anchor: Anchor = Anchor.CENTER
+    color: Union[np.ndarray, str] = 'cyan'
+    size: float = 12.0
+    blending: Blending = Blending.TRANSLUCENT
+    visible: bool = True
+
+    def __post_init__(self):
+        # private attributes not in dataclass
+        self._text_format_string: str = ''
+        self._values = None
+        self._mode: TextMode = TextMode.NONE
+        self.anchor = Anchor(self.anchor)
+        self.blending = Blending(self.blending)
+        with self.events.text.blocker():
+            self._set_text(self.text, self.n_text, self.properties)
 
     @property
     def values(self):
         """np.ndarray: the text values to be displayed"""
         return self._values
 
+    @property
+    def mode(self) -> str:
+        """str: The current text setting mode."""
+        return str(self._mode)
+
+    # Things that need to modify the value on set
+
+    def _on_translation_set(self, translation):
+        if not isinstance(translation, np.ndarray):
+            self.translation = np.asarray(translation)
+            return True
+
+    def _on_color_set(self, color):
+        if color != self.color:
```

Comment:
why have the `!=` catch here? seems a bit of "one off" to have it and unlike the others

Reply:
That's to avoid a `RecursionError` (comments [here](https://github.com/napari/napari/blob/0387244cea007d83aa1aa733dafe6f72f25e5fba/napari/utils/dataclass.py#L58-L61)).  There are other patterns we could employ, but since we have a setter calling another setter, one has to be careful.

Reply:
hmm yeah, all the more reason I'd prefer a different way to pass a single transformation function that get's called on setting for these simple cases.

Originally I thought that would block all event emission, but i now realize with the `super` call it wouldn't, it just prevents the recursion, so it makes sense, but requires some thought

Reply:
Thats what https://github.com/napari/napari/pull/1485#issuecomment-660565175  is proposing... (note, those types _are_ functions). Using color as an example, here are two I could support, let me know what you like:

1.  `Annotated` type ([PEP 593](https://www.python.org/dev/peps/pep-0593/)), that takes (up to) 3 arguments, with the first being what type(s) are accepted from the user, the second being a function that is called on the value when "getting" the value (can also be a callable `type`), and the third is the function that is called when "setting" the value.

```python
from typing_extensions import Annotated

# syntax = Annotated[type, get_func, set_func (if different than type)]

# color that may be either ndarray or string, and calls transform_color on set
color: Annotated[Union[np.ndarray, str], None, transform_color] = 'cyan'

# internal enum for an internal enum that "gets" as a string
anchor: Annotated[Anchor, str] = 'cyan'
```

2.  using `metadata` on [dataclass `fields`](https://docs.python.org/3/library/dataclasses.html#dataclasses.field)
```python

# where both get/set_as are optional. 
color: Union[np.ndarray, str] = field(default='cyan, metadata={'set_with': transform_color})

# example for an internal enum that "gets" as a string
anchor: Anchor = field(default=Anchor.CENTER, metadata={'get_with': str})
```

Reply:
@jni, if you're holding off on looking at this whole thing, i'd be curious to get your opinion on this one comment... (how to indicate some simple getter/setter transformation that used to live in @properties/setters)

Reply:
this is also all basically a "light" version of what we'd get with [pydantic](https://pydantic-docs.helpmanual.io/), (see https://github.com/napari/napari/issues/1436) which has a whole framework for expressing validators and type converters... so, with pydantic I _think_ it would just be:
```python
color: ColorType = 'cyan'
```
and pydantic guarantees that your class always has a valid `ColorType` (or an exception if it cannot convert)

Reply:
Yes, I was wondering if https://github.com/napari/napari/pull/1485#issuecomment-660565175 might be the answer!!

I really like the Annotated Type approach - get that that is a standard now, it's really I think what we want, and I imagine we could get most functionality we need with that triple.

I like the
```python
color: Annotated[Union[np.ndarray, str], None, transform_color] = 'cyan'
anchor: Annotated[Anchor, str] = 'cyan'
```

Syntax, I find that easy to read and understand. I prefer it to the `field`  syntax.

I guess the `ColorType` syntax is quite slick and is nice because we could reuse that in many places, but I still like the idea of growing into that slowly.

Not sure what your plan for this PR next is, but I'd be down for a version using the `Annotated` syntax, and then maybe all our custom get/ set would go away (I think we could even have a coerce translucent function for the blending case)!!!

Reply:
> Syntax, I find that easy to read and understand. I prefer it to the field syntax.

yeah, I also prefer it to the field syntax, except perhaps that i liked the `get_with`/`set_with` in the `field()` option... but we could do that with annotated too.  Which do you prefer?

```python
color: Annotated[ColorType, None, transform_color] = 'cyan'
# or 
color: Annotated[ColorType, {'set_with': transform_color}] = 'cyan'
```

> I guess the ColorType syntax is quite slick

we could probably do something like that internally without `pydantic` with a little type->converter `dict`.  I'll play with that.

> Not sure what your plan for this PR next is, but I'd be down for a version using the `Annotated` syntax

I'll go with that (but easy to switch if @jni hates it).  I think we just leave this PR open for a while, and I'll also make one for `ViewerModel` and stuff... until we get something that feels right for them all

Reply:
I guess I still lean to `color: Annotated[ColorType, None, transform_color] = 'cyan'`. It's slightly less explicit, but I think it will be such a major pattern we use as long as it is clearly documented somewhere then I think it's ok and it is cleaner to read.

> we could probably do something like that internally without pydantic with a little type->converter dict. I'll play with that.

yeah, but i'm also ok with us eventually going down the pydantic road too, so no need to spend too long.

>  I think we just leave this PR open for a while, and I'll also make one for ViewerModel and stuff... until we get something that feels right for them all

Ok sounds great. I'd like to do `Dims` just to get the experience working with the pattern too. Making the layer list version of the EVH really helped me understand it. I could do now, but might take both #1444 and #1475 to do it.

Reply:
> I'd like to do Dims just to get the experience working with the pattern too.

That would be awesome!  I'm keeping #1475 up to date as make changes here... and I think #1444 is mostly done/static at this point.  So I have no problem just making a new PR for each new object we're porting over, merging in #1475 and/or #1444, and then just letting those PRs sit there for a bit.  We can play with the pattern without worrying about having to revert.  If you hit a roadblock in `Dims`, just let me know and we'll decide if the dataclass needs to change, or if we just workaround it in `Dims`